nixos/security/wrappers: simplifications and a fix for #98863

robryk opened the pull request:
###### Description of changes

Simplification of SUID wrappers, which should make us care less about weird subtle properties of their environment and removal of assertions about those subtle properties. This (partially) fixes #98863: only for S[UG]ID binaries, so I'm also adding a regression test for that.

~~I didn't see any dedicated tests for the wrapper functionality, so I ran the `smokeping` test, which seems to be calling a SUID binary via its wrapper.~~ In the meantime, I've added a test for this functionality. TTBOMK (and, if I interpret the output correctly, according to `nixpkgs-review`) no packages are dependent on this code (only the NixOS activation scripts depend on wrappers).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

robryk committed:
nixos/security/wrappers: stop using `.real` files

Before this change it was crucial that nonprivileged users are unable to
create hardlinks to SUID wrappers, lest they be able to provide a
different `.real` file alongside. That was ensured by not providing a
location writable to them in the /run/wrappers tmpfs, (unless
disabled) by the fs.protected_hardlinks=1 sysctl, and by the explicit
own-path check in the wrapper. After this change, ensuring
that property is no longer important, and the check is most likely
redundant.

The simplification of expectations of the wrapper will make it
easier to remove some of the assertions in the wrapper (which currently
cause the wrapper to fail in no_new_privs environments, instead of
executing the target with non-elevated privileges).

Note that wrappers had to be copied (not symlinked) into /run/wrappers
due to the SUID/capability bits, and they couldn't be hard/softlinks of
each other due to those bits potentially differing. Thus, this change
doesn't increase the amount of memory used by /run/wrappers.

robryk commented:
FWIW I'm now looking at `make_caps_ambient` to see if it can be replaced with something that just looks at current capsets. If (as I suspect) that is infeasible, I will look into similarly building the capabilities that have to be elevated to ambient into the wrapper (which will also fix the TODO around `CAP_SETPCAP`), so that we can stop trying to look at own binary file at all. That will, I hope, make it clear that we don't need to make most of the assertions we currently do.

robryk commented:
@rnhmjoj Given that you modified this semantically last year, would you care to review?

rnhmjoj commented:
Yes, I did some cleanup on the module system part, but I'm not remotely qualified to touch security-related C code.
Calling @joachifm @xfix for this.

robryk commented on nixos/modules/security/wrappers/wrapper.nix:
```diff
@@ -8,6 +8,7 @@ stdenv.mkDerivation {
   hardeningEnable = [ "pie" ];
   CFLAGS = [
     ''-DWRAPPER_DIR="${parentWrapperDir}"''
+    ''-DSOURCE_PROG="${sourceProg}"''
```
Now I've looked at the other changes to wrappers around that time by the same fellow, and I can't see why that was done (at first I thought it was a prelude to deduplication, but deduplication wasn't done afterwards). Given the original author's lack of activity in nixos recently, would you consider it churlish to tag him and ask?

I'd like us to do try to investigate the reason (I'll try to spelunk more for it too; pointers are appreciated), but not to argue that an unknown reason must have existed and therefore continues to exist.

robryk replied:
https://github.com/NixOS/nixpkgs/pull/16654#discussion_r97961889 seems to be arguing for a single wrapper, so that it's more likely that it's cached. @7c6f434c do you still think that the wrapper being same for each suid/sgid/setcap binary is important?

robryk replied:
BTW. It might not have been obvious, but the "state before cca2e1155617304b5bcb3457309983cf8e99b067" was the state in a middle of a PR (d8ecd5eb0d30c3dc302e336755a4a1d6d0cb1ba4 comes from the same PR #16654, which wasn't squashed).

robryk commented:
Ping @Mic92 @7c6f434c @xfix. I would like to know at least whether there are any fundamental objections, because I want to make similar changes for passing in capabilities that should be made "inheritable" (well, ambient, because inheritable apparently doesn't do what we want).

Mic92 commented:
I think this solution is nicer with respect to usernamespaces. 
One idea to not increase the build time, what if the binary had a long string like `@@@@@@@@@@@@@@@@@@@@@@@@@@` as a template string which than gets truncated with `sed`?
Than we would not even need a gcc build for every wrapper. Otherwise this idea still looks good to me, I think we have to download stdenv with gcc and binutils most of the time anyway when building a nixos closure.

robryk commented:
> I think this solution is nicer with respect to usernamespaces.

Note that this doesn't fix anything in no_new_privs environments by itself, but I do intend this to be a first step of fixing that (next ones would be: doing the same to capabilities or reading capabilities from `/proc/self/exe` directly and removing most of the assertions about self and self's location, given that it's obvious that nothing relies on them then).

I would prefer not to try applying `sed` to ELF files: I would be worried that something I can't predict right now will change in the compiler's/linker's behaviour that will break it in a possibly-subtle way (I can't think of any concrete realistic scenario though, as long as the placeholder is unique enough). My two alternatives to rebuilding and to sed would be:
 - creative use of `objcopy` to place a constant in the binary (but that requires binutils, so might not be better at all),
 - placing the name of the target file in an xattr (which can be looked up by doing `getfattr(/proc/self/exe)`, which is guaranteed to point at the binary that's being executed).

rnhmjoj commented:
> I would prefer not to try applying sed to ELF files

Also, you usually have no more than a dozen wrappers and the program is quite short: I don't think this will slow down the build significantly. We could also try to take advantage of the binary cache for the most common wrappers (ping, sudo, etc.), if these are not cached already.

> placing the name of the target file in an xattr

Note that /run/wrappers is on a tmpfs partition. tmpfs has limited support for extended attributes: the "trusted" namespace is implemented but the attributes can be read by root only.

robryk commented:
> Also, you usually have no more than a dozen wrappers and the program is quite short: I don't think this will slow down the build significantly. We could also try to take advantage of the binary cache for the most common wrappers (ping, sudo, etc.), if these are not cached already.

IIUC @7c6f434c's concern from that old PR, it was about being able to add a new wrapper to a system without getting the whole toolchain fetched.

> Note that /run/wrappers is on a tmpfs partition. tmpfs has limited support for extended attributes: the "trusted" namespace is implemented but the attributes can be read by root only.

Thanks, TIL about tmpfs's restriction on xattrs (that `user.*` ones are not permitted). That does make the xattr approach something between "much worse" and "infeasible".

Mic92 commented:
I think that if we have nixos tests for the most common security wrappers than those should be cached already, so maybe probably we don't need to optimize for those cases.

robryk committed:
nixos/security/wrappers: read capabilities off /proc/self/exe directly

/proc/self/exe is a "fake" symlink. When it's opened, it always opens
the actual file that was execve()d in this process, even if the file was
deleted or renamed; if the file is no longer accessible from the current
chroot/mount namespace it will at the very worst fail and never open the
wrong file. Thus, we can make a much simpler argument that we're reading
capabilities off the correct file after this change (and that argument
doesn't rely on things such as protected_hardlinks being enabled, or no
users being able to write to /run/wrappers, or the verification that the
path readlink returns starts with /run/wrappers/).

robryk committed:
nixos/security/wrappers: remove all the assertions about readlink(/proc/self/exe)

Given that we are no longer inspecting the target of the /proc/self/exe
symlink, stop asserting that it has any properties. Remove the plumbing
for wrappersDir, which is no longer used.

Asserting that the binary is located in the specific place is no longer
necessary, because we don't rely on that location being writable only by
privileged entities (we used to rely on that when assuming that
readlink(/proc/self/exe) will continue to point at us and when assuming
that the `.real` file can be trusted).

Assertions about lack of write bits on the file were
IMO meaningless since inception: ignoring the Linux's refusal to honor
S[UG]ID bits on files-writeable-by-others, if someone could have
modified the wrapper in a way that preserved the capability or S?ID
bits, they could just remove this check.

Assertions about effective UID were IMO just harmful: if we were
executed without elevation, the caller would expect the result that
would cause in a wrapperless distro: the targets gets executed without
elevation. Due to lack of elevation, that cannot be used to abuse
privileges that the elevation would give.

This change partially fixes #98863 for S[UG]ID wrappers. The issue for
capability wrappers remains.

robryk committed:
nixos/security/wrappers: add one regression test for #98863

Note that this regression test checks only s[gu]id wrappers. The issue
for capability wrappers is not fixed yet.

Mic92 reviewed (approved):
I like how this just gets rid of a lot of source code.

robryk commented:
Great. In this case I've made all the subsequent changes necessary to fix #98863 for S[UG]ID binaries and added them to this PR (together with a regression test). Please take a look at commit descriptions to see motivations for the individual changes and arguments for their safety. I will try to fix things for binaries with file capabilities in a later PR, once I understand what invariants we'd want to preserve (e.g. what should happen if we can only assert some, but not all, relevant capabilities).

7c6f434c commented:
I think we have quite a few optional setuid wrappers even in NixOS options, so needing to fetch GCC for them is still a regression. Of course, for my system I need GCC anyway…

`/proc/self/exe` could be used by _concatenating_ the path to the executable, not replacing inside, but this assumes things about `/proc` (although if `/proc` is not mounted on NixOS, things are probably broken anyway?)

sed-replacing a string after checking it is indeed present and unique might be safe until linkers start implementing something very unexpected.

Mic92 commented:
> sed-replacing a string after checking it is indeed present and unique might be safe until linkers start implementing something very unexpected.

I don't think we need to support arbitrary linkers here and the ones I know don't do this.

Mic92 commented:
> I think we have quite a few optional setuid wrappers even in NixOS options, so needing to fetch GCC for them is still a regression. Of course, for my system I need GCC anyway…

However there are regression on both sides: Your regression is about disk space, the other one about correctness (see the assertions that make the wrappers crash in some circumstances).

robryk commented:
> /proc/self/exe could be used by concatenating the path to the executable, not replacing inside, but this assumes things about /proc (although if /proc is not mounted on NixOS, things are probably broken anyway?)

I don't understand this, can you be more verbose?

robryk commented:
> Your[7c6f434c's] regression is about disk space

My understanding was that it wasn't about disk space, but about ability to add a wrapped command on a system that doesn't have the whole C toolchain in its local store. I would also appreciate more verbosity here.

rnhmjoj commented:
Regarding tmpfs, I've found out the restriction to the "trusted" namespace is completely artificial: it's supposed to prevent a DoS attack from an unprivileged user attaching many large attributes to files on /tmp. The motivation seems very weak to me, if we agree on this we could remove the restriction with a trivial patch and continue with the xattr implementation.

7c6f434c commented:
Yes, it is not just HDD space (Nix is not that good at optimising that anyway), but also dependence on network availability for rebuilds. Or should we just explicitly make entire stdenv installed by default on NixOS?

Adding data after the end of an ELF file doesn't bother the kernel or the dynamic loader (e.g. SBCL does that to append the image to the base runtime). And as long as `/proc/self/exe` is available and readable (both can be false on some properly working GNU/Linux systems, but for wrappers on NixOS we have both conditions be true), you can read from there and not from `.real`; appending `\n/nix/store/…` to a binary wrapper is no harder than writing it to a separate file. Or, indeed, make an in-place replacement safe with some trick, which might be not worth it because then the trick needs to be maintained.

Mic92 commented:
I will try to come up with a simple linker script that would allow to append the path do the end of the file.

Mic92 commented:
Otherwise using objcopy from `binutils-unwrapped` should not add that much to the closure size.

7c6f434c commented:
I would think that appending to the end is just `(echo ; echo "$real") >> wrapper`? I think store path containing newlines would break our stdenv anyway…

7c6f434c commented:
Probably reopen, yes.

rnhmjoj commented:
So, do you not want to consider using extended attributes? That looks like the less hacky solution to me.

Mic92 commented:
I implemented reading `/proc/self/exe` to get the program name but it actually does not work for setgid binaries: https://github.com/NixOS/nixpkgs/commit/0c547bbba94637a3eaf9bfe4a55bd3a21eef71ea
They cannot open `/proc/self/exe`.

Mic92 commented:
> So, do you not want to consider using extended attributes? That looks like the less hacky solution to me.

Are you talking about doing a kernel patch here? This would break container support (i.e. systemd-nspawn) on non-nixos systems.

Mic92 commented:
I fixed it now! I added read permission for all users: ef5e4b43e51f7aa8e6fba9de43829d0c39dcc1ea
The only issue is that this is backwards-incompatible since it was not needed before. But maybe we can make it so it always is readable.

rnhmjoj commented:
> Are you talking about doing a kernel patch here

Yes, the patch would just be:
```
--- a/mm/shmem.c
+++ b/mm/shmem.c
@@ -2655,6 +2655,12 @@ static int shmem_xattr_handler_set(const struct xattr_handler *handler,
 	return simple_xattr_set(&info->xattrs, name, value, size, flags);
 }
 
+static const struct xattr_handler shmem_user_xattr_handler = {
+	.prefix = XATTR_USER_PREFIX,
+	.get = shmem_xattr_handler_get,
+	.set = shmem_xattr_handler_set,
+};
+
 static const struct xattr_handler shmem_security_xattr_handler = {
 	.prefix = XATTR_SECURITY_PREFIX,
 	.get = shmem_xattr_handler_get,
@@ -2672,6 +2678,7 @@ static const struct xattr_handler *shmem_xattr_handlers[] = {
 	&posix_acl_access_xattr_handler,
 	&posix_acl_default_xattr_handler,
 #endif
+	&shmem_user_xattr_handler,
 	&shmem_security_xattr_handler,
 	&shmem_trusted_xattr_handler,
 	NULL
```

> This would break container support (i.e. systemd-nspawn) on non-nixos systems.

So, does this mean NixOS has never required a kernel patch before? This is surprising to hear.

Mic92 commented:
> So, do you not want to consider using extended attributes? That looks like the less hacky solution to me.

It is actually not that hacky. To get the elf size, all you need is this formular:

```
  elf_header.e_shoff + (elf_header.e_shnum * elf_header.e_shentsize)
```

After that its just reading from this offset till the end of the file.

Mic92 commented:
> > Are you talking about doing a kernel patch here
> 
> Yes, the patch would just be:
> 
> ```
> --- a/mm/shmem.c
> +++ b/mm/shmem.c
> @@ -2655,6 +2655,12 @@ static int shmem_xattr_handler_set(const struct xattr_handler *handler,
>  	return simple_xattr_set(&info->xattrs, name, value, size, flags);
>  }
>  
> +static const struct xattr_handler shmem_user_xattr_handler = {
> +	.prefix = XATTR_USER_PREFIX,
> +	.get = shmem_xattr_handler_get,
> +	.set = shmem_xattr_handler_set,
> +};
> +
>  static const struct xattr_handler shmem_security_xattr_handler = {
>  	.prefix = XATTR_SECURITY_PREFIX,
>  	.get = shmem_xattr_handler_get,
> @@ -2672,6 +2678,7 @@ static const struct xattr_handler *shmem_xattr_handlers[] = {
>  	&posix_acl_access_xattr_handler,
>  	&posix_acl_default_xattr_handler,
>  #endif
> +	&shmem_user_xattr_handler,
>  	&shmem_security_xattr_handler,
>  	&shmem_trusted_xattr_handler,
>  	NULL
> ```
> 
> > This would break container support (i.e. systemd-nspawn) on non-nixos systems.
> 
> So, does this mean NixOS has never required a kernel patch before? This is surprising to hear.

I am using it without patches for quite a while without any issues. Patches are only needed for stuff like modprobe, but this is not of concern for containers anyway.

Mic92 commented:
`Cannot open /proc/self/exe: Permission denied` 

> I implemented reading `/proc/self/exe` to get the program name but it actually does not work for setgid binaries: [0c547bb](https://github.com/NixOS/nixpkgs/commit/0c547bbba94637a3eaf9bfe4a55bd3a21eef71ea) They cannot open `/proc/self/exe`.
> Ouch. OK, I am stupid for not understanding this problem (not everyone is allowed to read /proc/self/exe) does indeed happen (without strange configuration, just on Nixpkgs kernels)…

It's actually just a configuration issue on our side. We currently only allow owner and not groups to read, but I got it fixed.
I opened up a new PR here: https://github.com/NixOS/nixpkgs/pull/201536

Mic92 commented:
> > The only issue is that this is backwards-incompatible since it was not needed before. But maybe we can make it so it always is readable.
> 
> Well, the wrappers are buit from a public source so not supposed to be secret, and «make them readable» happens at the same time as new wrapper gets used, so each generation will always have these things in a compatible way?

Yeah I was not worried about leaking secrets here, but more about people using different options for the `permissions` option of the wrapper, but I got this fixed now as well.

robryk commented:
Note that we don't need to parse elf at all: we can simply append a string
and then append its length (and then start by reading the length).

I don't really understand the precise requirement of "being able to rebuild
nixos workout network access" and I expect that random other changes could
break it without anyone realizing. I would like us to at the very least
have a description of what is the property we want to preserve (and at best
tests that check that property) before we do things whose only goal is to
preserve that property.

Jörg Thalheim ***@***.***> schrieb am Mi., 16. Nov. 2022,
20:14:

> The only issue is that this is backwards-incompatible since it was not
> needed before. But maybe we can make it so it always is readable.
>
> Well, the wrappers are buit from a public source so not supposed to be
> secret, and «make them readable» happens at the same time as new wrapper
> gets used, so each generation will always have these things in a compatible
> way?
>
> Yeah I was not worried about leaking secrets here, but more about people
> using different options for the permissions option of the wrapper, but I
> got this fixed now as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/199599#issuecomment-1317541258>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALHHI3RYU3UCUOWHZEIUYDWIUXDFANCNFSM6AAAAAARXUWF4M>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Mic92 commented:
> Note that we don't need to parse elf at all: we can simply append a string and then append its length (and then start by reading the length). I don't really understand the precise requirement of "being able to rebuild nixos workout network access" and I expect that random other changes could break it without anyone realizing. I would like us to at the very least have a description of what is the property we want to preserve (and at best tests that check that property) before we do things whose only goal is to preserve that property.

If we append the length the end or have the length from the elf file does not make a big difference here, does it?
`being able to rebuild nixos workout network access`: It means that one does not need to download any packages to install a basic nixos (i.e. gcc/binutils) but instead it works from the installer itself. But I think in general the idea was that the installation closure should not also include a full stdenv with all its tools.

robryk commented:
> If we append the length the end or have the length from the elf file does not make a big difference here, does it? In fact reading backwards seems a bit more complex as it requires some seeks.

It doesn't require us to use elf parsing libraries at all (which is _not_ relevant security-wise here at all, but which I always intuitively see as a small plus due to their complexity), and if we want to have more than one piece of data transferred that way we'll need to do something that's similarly hard anyway. I would like to use a similar mechanism to provide the capabilities to be set, so that we don't do the terribly weird special casing of setpcap that prevents anyone from ever requesting it, which will require one more piece of data there.

>  being able to rebuild nixos workout network access: It means that one does not need to download any packages to install a basic nixos (i.e. gcc/binutils) but instead it works from the installer itself. But I think in general the idea was that the installation closure should not also include a full stdenv with all its tools.

I'm confused how that would be relevant _for this change_:
 - if basic nixos is "nixos with configuration.nix generated by `nixos-generate-config`" then this will trivially preserve that property: if we had the single wrapper derivation's result included, we'll have all the standard wrapper derivations included,
 - if basic nixos is "nixos with ~any configuration.nix", then this requires stdenv, because one can e.g. set overrides in configuration.nix that cause packages to be built in versions that are surely not cached.

robryk commented:
> Not for the current version; it was me whining about compiling afresh a new wrapper for each target.

I don't see how that's relevantly changed _by the version that compiles a new wrapper for each target._

7c6f434c commented:
Well, in a _good_ situation if the set of packages implied by your configuration does not change, you can just rebuild offline and change whatever configs. It is broken from time to time, sometimes leading to complaints. «I should make this command setuid after all» is a localised configuration change, not a change to the packages installed or indirectly used. So it's nicer if this change can be done on the go.

delroth commented:
> Well, in a good situation if the set of packages implied by your configuration does not change, you can just rebuild offline and change whatever configs. It is broken from time to time, sometimes leading to complaints.

If this is a property we want to uphold for nixpkgs/NixOS, then it needs to be properly formulated and documented as a policy. I don't think it's fair to try and push what is basically your *preference* as a reviewer to new contributors trying to improve the state of a mostly unmaintained (but critical) part of NixOS.

FWIW this is the first time I hear anyone complaining about having to download stdenv as part of a config rebuild... there's so many things on NixOS that end up fetching stdenv randomly that it's basically cached anywhere I use Nix tooling regularly.

(Also, if we have any NixOS module that registers SUID wrappers but also don't have NixOS tests being built by Hydra... that sounds to me like a testing gap that should be addressed.)

robryk commented:
I'm not sure what is the store in the installer prepopulated with. If it had all the build dependencies of its contents, this would not be a problem -- it would always have stdenv in it. If it has only runtime dependencies, then we are already in a lot of trouble: `grep '\.override' nixos -R` in nixpkgs shows many places where a config entry is used to override a parameter of a derivation that build from sources that need to be fetched (and which would be absent from the store in that scenario), regardless of the dependency on stdenv.

robryk commented:
This PR is hanging in a state where one person expressed concerns, and eventually stopped responding to requests to make those concerns legible, so that they can be taken into account for all changes. I don't know whether this is a manifestation of an unstated consensus that we don't want this change and I should just abandon it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/where-did-you-get-stuck-in-the-nix-ecosystem-tell-me-your-story/31415/44

xfix commented:
I think this change makes sense, but I'm also not confident verifying a security change like this (all I did was adding `argc >= 1` check). Most of those checks probably don't really add value, but setuid programming is tricky...

7c6f434c commented:
There is _never_ a consensus on _anything_. 

I was under impression that here either there would be an update to make wrappers readable and just read the path from the end, which would be better, or you would state explicitly that you were not going to do that — as your comments on that looked still in doubt — and then I guess @Mic92 would merge it as the committer who likes the current idea… and as I thought there would be an update one way or another and for me personally current version is a negative (well if someone merges it so be it but it won't be me merging), I didn't actively check what is going on when it didn't show up again in notifications.

delroth commented:
> I was under impression that here either there would be an update to make wrappers readable and just read the path from the end, which would be better

FWIW I personally disagree that this would be better. It trades off "needing a C compiler" for "extra complexity in a security critical component". And especially when it appears there's no clear maintenance/ownership story for said component, I think making it as simple as possible to grasp and audit is the right direction.

> personally current version is a negative (well if someone merges it so be it but it won't be me merging)

I'd interpreted your previous comments on this PR as a "I veto merging this in its current state", so I think this clarification helps a lot!

delroth reviewed (approved):
- Any simplification to a security-critical component is imo welcome. Especially, making it not rely on FS lookups as much is great, since these are notoriously difficult to get right (many levers to manipulate as an unprivileged user to try and trick the wrapper, and very race-prone).
- I've spent some time staring at the wrapper diff and am reasonably convinced that it doesn't introduce any new vulnerability. All the checks removed seem to be made redundant by the preprocessor-hardcoding of `SOURCE_PROG`.
- Fixes a (valid) reported bug and adds new regression tests. Existing tests all still pass.
- The requirement for having a C compiler to build new wrappers has been discussed a lot on this PR. I can understand differences in opinion here, but I think @robryk has made a good job justifying that in practice the effects are likely to be roughly inexistant. Summarizing: there are many things one can do in their NixOS configuration that can end up requiring a C compiler (overrides, incl. some that already exist via NixOS modules options), and any common SUID wrapper should already be getting built by Hydra via NixOS tests. In absence of a well defined policy, I don't think we should apply extra arbitrary standards here without a strong motive.

I'll give a few more days for anyone to veto this, otherwise happy to press the merge button myself.

fabianhjr commented:
I think this is causing a build issue (at least on staging-next)

```
building Nix...
building the system configuration...
error:
       … while calling the 'head' builtin

         at /home/fabian/Development/Open/nixpkgs/main/lib/attrsets.nix:820:11:

          819|         || pred here (elemAt values 1) (head values) then
          820|           head values
             |           ^
          821|         else

       … while evaluating the attribute 'value'

         at /home/fabian/Development/Open/nixpkgs/main/lib/modules.nix:800:9:

          799|     in warnDeprecation opt //
          800|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          801|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: cannot convert a function to JSON

       at /home/fabian/Development/Open/nixpkgs/main/nixos/modules/security/wrappers/default.nix:8:21:

            7|
            8|   securityWrapper = sourceProg : pkgs.callPackage ./wrapper.nix {
             |                     ^
            9|     inherit sourceProg;
```

Investigating a possible fix.

delroth commented:
This seems to be:

```
    security.apparmor.includes."nixos/security.wrappers" = ''
      include "${pkgs.apparmorRulesFromClosure { name="security.wrappers"; } [
        securityWrapper
      ]}"
    '';
```

I have AppArmor disabled on all my systems so my personal testing didn't show this :/

Will revert for now and let @robryk do a respin on this. Thanks for the report.

delroth commented:
Reverted in #251095.

robryk commented:
New attempt in #251770

SuperSandro2000 commented:
> I have AppArmor disabled on all my systems so my personal testing didn't show this :/

that's because it is disabled by default

delroth commented:
The hardened profile enables it IIRC, so it's probably not a super rare configuration.

SuperSandro2000 commented:
In my experience the hardened profile is super rare to find in the wild. For example the hardened kernel didn't compile for multiple months on unstable in the past and no one seemed to really care.
